src/test/solution_space/solution_graph/graph_vertices_dist_tests.py

```diff
@@ -0,0 +1,186 @@
+# (Seems GumTreeDiff distance is not commutative :c )
+anon_distance = [[0, 8, 12, 20, 33, 21],
+                 [8, 0, 16, 20, 30, 21],
+                 [12, 16, 0, 28, 38, 27],
+                 [20, 20, 28, 0, 11, 16],
+                 [33, 33, 38, 11, 0, 24],
+                 [24, 21, 27, 16, 24, 0]]
+
+# Distances between all canon trees for all 3 vertices:
+canon_distance = {VERTEX.VERTEX_0: {VERTEX.VERTEX_0: 0, VERTEX.VERTEX_1: 18, VERTEX.VERTEX_2: 19},
+                  VERTEX.VERTEX_1: {VERTEX.VERTEX_0: 18, VERTEX.VERTEX_1: 0, VERTEX.VERTEX_2: 16},
+                  VERTEX.VERTEX_2: {VERTEX.VERTEX_0: 19, VERTEX.VERTEX_1: 16, VERTEX.VERTEX_2: 0}}
+
+
+# If we add fragments to the solution graph in that order, vertices will be created or updated like this:
+#  *adding fragment_0*  vertex_0 created
+#  *adding fragment_3*  vertex_1 created
+#  *adding fragment_1*  vertex_0 updated
+#  *adding fragment_5*  vertex_2 created
+#  *adding fragment_2*  vertex_0 updated
+#  *adding fragment_4*  vertex_1 updated
+fragment_indices_to_add = [0, 3, 1, 5, 2, 4]
+
+
+# Finds distance matrix between all vertices in given dict, using anon trees which indices listed in the dict.
+#
+# For example, for added_indices_by_vertex: {VERTEX.VERTEX_0: [0, 3],
+#                                            VERTEX.VERTEX_1: [4]}
+# it will find all distances between
+# 1. VERTEX_0 with fragment_0, fragment_3
+# 2. VERTEX_1 with fragment_4.
+#
+# It's useful for finding intermediate distance matrix while adding new fragments.
+def find_dist_matrix(added_indices_by_vertex: Dict[VERTEX, List[int]]) -> List[List[int]]:
+    dist_matrix = []
+    for i, src_vertex in enumerate(added_indices_by_vertex.keys()):
+        dist_matrix.append([])
+        src_indices = added_indices_by_vertex[src_vertex]
+        for j, dst_vertex in enumerate(added_indices_by_vertex.keys()):
+            dst_indices = added_indices_by_vertex[dst_vertex]
+            canon_dist = canon_distance[src_vertex][dst_vertex]
+            anon_dist = min([anon_distance[k][l] for k in src_indices for l in dst_indices])
+            dist_matrix[i].append(min(anon_dist, canon_dist))
+    return dist_matrix
+
+
+def get_code_info_chain(sources: List[str]) -> List[Tuple[Code, CodeInfo]]:
+    user = User()
+    return [(create_code_from_source(s), CodeInfo(user)) for s in sources]
+
+
+@pytest.mark.skipif(to_skip(current_module_level=TEST_LEVEL.SOLUTION_SPACE), reason=TEST_LEVEL.SOLUTION_SPACE.value)
+class TestDistBetweenVertices():
+
+    # Check all fragments in same vertices have the same canon trees
+    @pytest.mark.parametrize('vertex', [v for v in VERTEX])
+    def test_same_canon_trees_in_same_vertices(self, vertex: VERTEX):
+        same_canon_fragments = [all_fragments[i] for i in INDICES_BY_VERTEX[vertex]]
+        canon_trees = [get_trees(f, {TREE_TYPE.CANON})[0] for f in same_canon_fragments]
+        for canon_tree_1, canon_tree_2 in itertools.product(canon_trees, repeat=2):
```

Comment:
Please use [subtests](https://github.com/pytest-dev/pytest-subtests) for separating tests. You can also see an example in the `diff_handler_test.py`